src/main/clojure/jarl/builtins/comparison.clj

```diff
@@ -0,0 +1,116 @@
+(ns jarl.builtins.comparison
+  (:require [jarl.builtins.utils :refer [type-sort-order]]))
+
+; From the OPA go docs on ast.Compare:
+;
+; If a is less than b, the return value is negative. If a is greater than b,
+; the return value is positive. If a is equal to b, the return value is zero.
+;
+; Different types are never equal to each other. For comparison purposes, types
+; are sorted as follows:
+;
+; nil < Null < Boolean < Number < String < Var < Ref < Array < Object < Set <
+; ArrayComprehension < ObjectComprehension < SetComprehension < Expr < SomeDecl
+; < With < Body < Rule < Import < Package < Module.
+
+(declare rego-compare)
+
+(defn- first-diff-pair [a b]
+  (let [pairs (partition 2 (interleave a b))
+        diff-pairs (filter #(not= (first %) (second %)) pairs)]
+    (first diff-pairs)))
+
+; Arrays and Refs are equal if and only if both a and b have the same length
+; and all corresponding elements are equal. If one element is not equal, the
+; return value is the same as for the first differing element. If all elements
+; are equal but a and b have different lengths, the shorter is considered less
+; than the other.
+(defn- vector-compare [a b]
+  (let [ca (count a)
+        cb (count b)]
+    (if (= ca cb)
+      (let [[x y] (first-diff-pair a b)]
+        (rego-compare x y))
+      (let [[x y] (first-diff-pair a b)]
+        (if (nil? x)
+          (compare ca cb)
+          (rego-compare x y))))))
+
+; Objects are considered equal if and only if both a and b have the same sorted
+; (key, value) pairs and are of the same length. Other comparisons are
+; consistent but not defined.
+(defn- map-compare [a b]
+  (let [sa (into (sorted-map) a)
+        sb (into (sorted-map) b)
+        keys-comp (vector-compare (keys sa) (keys sb))]
+    (if-not (zero? keys-comp)
+      keys-comp
+      (let [values-comp (vector-compare (vals sa) (vals sb))]
+        (if (zero? values-comp)
```

Comment:
If `keys-comp==0` and `values-comp==0`, under what circumstances will this further logic be needed?

Reply:
I see: sizes can still differ. Could the size check be rolled into `vector-compare`, or does that muddy the waters?

Reply:
Yeah, I think the size check is performed there already, so this probably wouldn't come back as zero.. just not sure how much each function should rely on the internals of the others. I'll leave this for now.